test/box/net.box_closing_without_lost_gh-6338.test.lua

```diff
@@ -0,0 +1,111 @@
+net_box = require('net.box')
+fiber = require('fiber')
+test_run = require('test_run').new()
+
+test_run:cmd("setopt delimiter ';'")
+function connection_send_async_requests(count)
+    local conn = net_box.connect(box.cfg.listen)
+    local space = conn.space.test
+    for i = 1, count do space:replace({i}, {is_async = true}) end
+    conn:close()
+    test_run:wait_cond(function() return box.stat.net().CONNECTIONS.current == 0 end)
+end;
+function create_space()
+    local s = box.schema.space.create('test')
+    local _ = s:create_index('primary')
+    return s
+end;
+function check_requests_result_and_truncate_space(space, count)
+    local rc = space:select()
+    space:truncate()
+    return (#rc == count)
+end;
+test_run:cmd("setopt delimiter ''");
+
+box.schema.user.grant('guest', 'read,write,execute,create,drop', 'universe',  nil, {if_not_exists = true})
+
+s = create_space()
+connection_send_async_requests(box.cfg.net_msg_max - 1, true)
+assert(check_requests_result_and_truncate_space(s, box.cfg.net_msg_max - 1))
+
+errinj = box.error.injection
+-- A special test to check that there is no hang when the connection is
+-- closed from the server side, when not all requests have been sent yet.
+-- (see send_cond:broadcast() usage from net_box.lua)
+cond = fiber.cond()
+channel = fiber.channel(1)
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    errinj.set("ERRINJ_NETBOX_IO_ERROR", true)
+    errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+-- We are sending a very large number of requests here to ensure that
+-- they will not have time to send by the time the connection is closed.
+conn = net_box.connect(box.cfg.listen)
+space = conn.space.test
+errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
+_ = space:replace({1}, {is_async = true})
+test_run:cmd("setopt delimiter ';'")
+cond:broadcast()
+conn:close()
+test_run:cmd("setopt delimiter ''");
+channel:get()
+errinj.set("ERRINJ_NETBOX_IO_ERROR", false)
+
+-- A special test to check that async request discards, if connection is
+-- closing and waits until send_buf became empty.
+space = nil
+not_ok = nil
+rc = nil
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    not_ok, rc = pcall(
+        space.replace, space, {1}, {is_async = true}
+    )
+    errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+conn = net_box.connect(box.cfg.listen)
+space = conn.space.test
+errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
+_ = space:replace({1}, {is_async = true})
+test_run:cmd("setopt delimiter ';'")
+cond:broadcast()
+conn:close()
+test_run:cmd("setopt delimiter ''");
+assert(not not_ok)
+-- new error: Connection is closing
+assert(rc)
```

Comment:
Top-level assertions are pointless in a diff test.

Reply:
Fixed, apply new version of test